distributed/diagnostics/memray.py

```diff
@@ -0,0 +1,165 @@
+
+                int: Number of workers to fetch concurrently
+                True: All workers concurrently
+                False: One worker at a time
+
+        **memray_kwargs
+            Keyword arguments to be passed to memray.Tracker, e.g.
+            {"native_traces": True}
+        """
+        self.directory = directory
+        self.client = get_client()
+        self.memray_kwargs = memray_kwargs
+        self.report_args = report_args
+        self.fetch_reports_parallel = fetch_reports_parallel
+        if not workers or isinstance(workers, int):
+            workers_returned = sorted(self.client.run(lambda: None))
+            nworkers = len(workers_returned)
+            if isinstance(workers, int):
+                nworkers = workers
+            workers = workers_returned[:nworkers]
+        self.workers = list(workers)
+        self.filename = uuid.uuid4().hex
+
+    def __enter__(self) -> None:
+        assert all(
+            self.client.run(
+                _run_memray, filename=self.filename, **self.memray_kwargs
+            ).values()
+        )
+        # Sleep for a brief moment such that we get
+        # a clear profiling signal when everything starts
+        time.sleep(0.1)
+
+    def __exit__(self, *args: Any) -> None:
+        os.makedirs(self.directory, exist_ok=True)
+
+        client = get_client()
+        if self.fetch_reports_parallel is True:
+            fetch_parallel = len(self.workers)
+        elif self.fetch_reports_parallel is False:
+            fetch_parallel = 1
+        else:
+            fetch_parallel = self.fetch_reports_parallel
+
+        for w in partition(fetch_parallel, self.workers):
+            try:
+                profiles = client.run(
+                    _fetch_profile,
+                    filename=self.filename,
+                    report_args=self.report_args,
+                    workers=w,
+                )
+                for worker_addr, profile in profiles.items():
+                    path = f"{self.directory}/{worker_addr.replace('/','')}.html"
+                    with open(
+                        path,
+                        "wb",
+                    ) as fd:
+                        fd.write(profile)
+
+            except Exception:
+                logger.exception(
+                    "Exception during report downloading from worker %s", w
+                )
+
+
+def _run_memray(dask_worker: Worker, filename: str, **kwargs: Any) -> bool:
+    if hasattr(dask_worker, "_memray"):
+        dask_worker._memray.close()
+
+    if os.path.exists(filename):
+        os.remove(filename)
+
+    dask_worker._memray = contextlib.ExitStack()  # type: ignore[attr-defined]
+    dask_worker._memray.enter_context(  # type: ignore[attr-defined]
+        memray.Tracker(filename, native_traces=True, **kwargs)
+    )
+
+    return True
+
+
+def _fetch_profile(
+    dask_worker: Worker, filename: str, report_args: Sequence[str] | Literal[False]
```

Comment:
Can we add a small docstring here? The function has side-effects (file-generation), would be nice to mention that

Reply:
Well, this is an internal 20line function. I don't think we need a doc string for this, do we?

Reply:
Yeah sorry, agree that a proper docstring is too much. I had something like

```
"""Creates the memray report files and fetches the results"""
```

in mind